[pvr] finally make it possible to use multiple PVR clients

Jalle19 opened the pull request:
I've grown tired of telling users that it's not possible to use more than one addon at a time due to a bug, so this is my attempt at fixing it. We really need a solution for this for Isengard so I'd appreciate it if we can keep the bikeshedding to a minimum and focus on getting something that works done, even though the solution might not be the prettiest.

The problem:
- the addons are started by `CPVRClients` on a background thread. Meanwhile, `CPVRManager` used a simple sleep loop while waiting for at least one addon to load before continuing. If the second (or third) addon didn't load during the same 50 millisecond sleep duration, the startup would continue with just one addon.

The solution:
- use a boolean in `CPVRClients` to indicate whether all clients have been loaded or not. From `CPVRManager` we call a method which blocks until this boolean's condition variable is signaled.

Ping @opdenkamp @ksooo @xhaggi @FernetMenta

I'm not convinced this is the ultimate solution, but at least it works and it doesn't introduce arbitrary sleep durations.

Jalle19 committed:
[cosmetics] only check m_addons for NULL once

Jalle19 committed:
[pvr] remove bInitialiseAllClients boolean parameter

Jalle19 committed:
[pvr] add a way for CPVRClients to signal that it has loaded (or at
least tried to load) all enabled addons. This fixes the ancient bug
where it's unreliable to use more than one PVR client simultaneously.

Closes #14498

xhaggi commented:
does not runtime test it, but looks good.

ksooo commented:
Looking good, wrt pragmatic approach to improve things for Isengard. Though I did not runtime test.

Jalle19 commented:
It's easy to runtime test, just enable pvr.demo and start Kodi a few times, checking that both your normal PVR client and the demo client gets loaded.

xhaggi commented:
IMO shove this in, even better than it was before.

Jalle19 commented:
I'd still like to hear @opdenkamp's thoughts

MartijnKaijser commented:
I don't see this as a fix

Jalle19 commented:
@MartijnKaijser why not? We have always supported multiple clients, it just hasn't ever really worked that well so we have always adviced people to not do it.

xhaggi commented:
totally agree with @Jalle19, it's a feature since the beginning but it didn't work in some cases. This will fix it and therefore it's a fix ;)

freddysmith1 commented:
For some people this wont bother them in anyway but those who have a need for it will be very greatful   their is people who use two different backends

xhaggi commented:
@Jalle19 i would say let's shove it in for Beta1 so we have a wide range of testings till our final release.

jenkins build this please.

Jalle19 commented:
I agree, any breakage should be easily spotted.

ksooo commented:
:+1: for inclusion in beta
:+1: for being a fix

freddysmith1 commented:
Any rough estimate when this will be included in a nightly build?

xhaggi commented:
any objections, if not i'll merge it in later today.

FernetMenta commented:
i am traveling and would like to have a closer look at this.

opdenkamp commented:
and I've been out of the loop for a couple of days because I was ill and 
haven't looked at it yet either.

On 29-04-15 13:41, Rainer Hochecker wrote:
> i am traveling and would like to have a closer look at this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7020#issuecomment-97397823>.
>

FernetMenta commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -1159,9 +1161,22 @@ bool CPVRClients::UpdateAndInitialiseClients(bool bInitialiseAllClients /* = fal
+  // indicate that addons have now been initialised
+  {
+    CSingleLock lock(m_critSection);
+    m_bAddonsInitialised = true;
```
what is the purpose of m_bAddonsInitialised? it is evaluated nowhere

Jalle19 replied:
Ah, good catch, the wait() is supposed to use it. I guess it works without it too, but it's not very obvious.

FernetMenta replied:
using low level objects like condition is not good either. we have CEvent for this kind of jobs

Jalle19 commented:
@FernetMenta I changed it to use an event instead and removed the superfluous boolean. Tested by injecting sleeps here and there and it seems to work just fine.

FernetMenta commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -1159,9 +1160,17 @@ bool CPVRClients::UpdateAndInitialiseClients(bool bInitialiseAllClients /* = fal
 
+void CPVRClients::WaitForInitialisation()
+{
+  m_loadedEvent.Wait();
```
A timeout for wait prevents from waiting forever in case something goes wrong.

opdenkamp replied:
shouldn't wait for all add-ons to be loaded forever anyway. and you need to signal the event when the pvrmanager is stopped, or this will deadlock

Jalle19 replied:
What would be a reasonable timeout? 1 second?

opdenkamp commented on xbmc/pvr/PVRManager.cpp:
```diff
@@ -627,15 +627,16 @@ bool CPVRManager::Load(void)
 
-  /* load at least one client */
-  while (IsInitialising() && m_addons && !m_addons->HasConnectedClients())
-    Sleep(50);
```
you really don't want to wait for all add-ons to be initialised and connected, but should continue to load when one of them is initialised, and add new data if there are more add-ons to be initialised. with this code, one add-on with a failing connection (e.g. backend offline) will block pvr from being loaded completely.

Jalle19 replied:
Since the focus of this PR is to get something in that works, can we agree on a suitable timeout and focus on refactoring this properly some time in the future?

opdenkamp replied:
The focus is to fix this properly, not to shove something in at the last moment that only works half.


-------- Oorspronkelijk bericht --------
Van: Sam Stenvall <notifications@github.com> 
Datum:30-04-2015  11:42  (GMT+01:00) 
Aan: xbmc/xbmc <xbmc@noreply.github.com> 
Cc: Lars Op den Kamp <opdenkamp@gmail.com> 
Onderwerp: Re: [xbmc] [pvr] finally make it possible to use multiple PVR clients
  (#7020) 

In xbmc/pvr/PVRManager.cpp:

>  
> -  /* load at least one client */
> -  while (IsInitialising() && m_addons && !m_addons->HasConnectedClients())
> -    Sleep(50);
Since the focus of this PR is to get something in that works, can we agree on a suitable timeout and focus on refactoring this properly some time in the future?

—
Reply to this email directly or view it on GitHub.

Jalle19 replied:
I specifically stated in the PR description that I know this is not the best way but that we need some kind of fix for the next release.

opdenkamp replied:
why is this so urgent that it needs a half fix that may or may not be 
fixed properly at some point? you're removing the "one add-on loaded" 
code now, which was there on purpose, which used to work a couple of 
releases back and which was broken while refactoring a while ago.

unless there's a very good reason to not fix this properly now 
(shouldn't be that much work), it gets a -1 from me at this point in the 
release schedule.

On 30-04-15 12:26, Sam Stenvall wrote:
> In xbmc/pvr/PVRManager.cpp
> <https://github.com/xbmc/xbmc/pull/7020#discussion_r29418339>:
>
>>
>> -  /* load at least one client */
>> -  while (IsInitialising() && m_addons && !m_addons->HasConnectedClients())
>> -    Sleep(50);
>
> I specifically stated in the PR description that I know this is not the
> best way but that we need some kind of fix for the next release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7020/files#r29418339>.
>

freddysmith1 commented:
Any sort of fix is better then no fix at all

opdenkamp commented:
it's not.

On 30-04-15 12:32, freddysmith1 wrote:
> Any sort of fix is better then no fix at all
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7020#issuecomment-97732675>.
>

Jalle19 committed:
[pvr] improve logging of client load statuses during startup

Jalle19 commented:
> why is this so urgent that it needs a half fix that may or may not be fixed properly at some point? you're removing the "one add-on loaded" code now, which was there on purpose, which used to work a couple of releases back and which was broken while refactoring a while ago

It has been broken for over two years and I don't see anyone else fixing it.

> unless there's a very good reason to not fix this properly now (shouldn't be that much work)

Do you have any suggestions on how this should be done properly? Either way I've updated the PR to wait a maximum of 5 seconds instead of forever.

Jalle19 commented:
What's the verdict on this?

opdenkamp commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -47,6 +47,8 @@ using namespace EPG;
 /** sleep time in milliseconds when no auto-configured add-ons were found */
 #define PVR_CLIENT_AVAHI_SLEEP_TIME_MS     (250)
 
+const int CPVRClients::CLIENT_LOAD_TIMEOUT = 5000;
```
just #define like the line above please, and add _MS to the name please and a comment. it's only used within this class anyway

opdenkamp commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -176,6 +178,7 @@ void CPVRClients::Unload(void)
   m_strPlayingClientName = "";
 
   m_clientMap.clear();
+  m_loadedEvent.Reset();
```
does this signal the waiting thread? haven't used this high level event implementation much

opdenkamp replied:
from the looks of it it doesn't, which means that we can block up to 5 seconds now on exit or when restarting the pvr manager, not good

Jalle19 replied:
So we should call `Set()` before reset?

Jalle19 replied:
@FernetMenta do you know how this `CEvent` is supposed to be used?

FernetMenta replied:
You have instantiated an automatic event. Means that wait resets it. You should reset it when starting the thread here and signal here on exit. IMO you can reduce the timeout as well. It just spins another loop if it times out, right?

Jalle19 replied:
The loop is gone so the timeout is absolute.

FernetMenta replied:
hmm, why is it gone. now you would kill a slow but vital init process in case of timeout.

opdenkamp commented:
we should not wait for all clients to be connected, as explained earlier. that's why the code was structured like this.

it should just return from the method when 1 add-on connected, and the updater thread for the clients should keep trying to connect to the other add-ons (if any) while they're not all connected.

Jalle19 commented:
There's much work to be done to implement it the way you're suggesting:

a) `m_channelGroups->Load()` are only called once (when the manager starts), if all clients are not loaded by then their "content" will not be loaded regardless of whether the client gets loaded eventually.

b) once the channel group container has started to load it will search the database and find all channels from the client that didn't start on time and delete them. This can take a lot of time. Same thing happens with the EPG data.

c) even if we'd change the code so that the channel group container etc. is reloaded when a new client is connected there'd be a ton of pointless database operations in between.

I don't see why we can't go with the current solution for now, the absolutely worst case scenario is that the "Loading channels from clients" dialog is shown for five seconds during startup, alternatively during shutdown if the user quits Kodi as soon as possible after launching it.

Jalle19 commented:
@FernetMenta I updated the event handling, is this how you meant?

opdenkamp commented:
The main advantage of having multiple pvr clients imo is to be able to switch between backends easily. All the rest can be consolidated into 1 backend. The code originally did handle this correctly a couple of years ago, but it broke with all the refactoring. Rather than doing it like this, and removing the parts of the code that were structured to handle it properly in the process, the points you mentioned should be fixed. It's one of the main reasons for having the updater threads in pvr...

Jalle19 commented:
@opdenkamp since you're pretty much the only one who is familiar with how it should work (I didn't use PVR many years ago), any chance you might look into it at some point?

Jalle19 commented:
@opdenkamp while thinking about this again, I realized that the way you want it to work will make the startup code incredibly complicated. Think of this scenario:

1. The user normally has two addons enabled so he has channels and guide data from both addons.
2. Start Kodi, and only one addon gets started "in time".
3. All the channels and guide data from the "stale" addon is removed. This actually takes quite some time, especially the guide data.
4. Suddenly the second addon wakes up and starts to add data, all the while the data is still being deleted. This sounds like a recipe for a disaster.

AlwinEsch commented on xbmc/pvr/PVRManager.cpp:
```diff
@@ -627,15 +627,28 @@ bool CPVRManager::Load(void)
+  {
+    CLog::Log(LOGINFO, "PVRManager - %s - %d of %d enabled clients connected in time, continue to start", 
+      __FUNCTION__, numConnectedClients, numEnabledClients);
+  }
```
Not sure what currently is the status of this request, but is there something on "numConnectedClients" check missing? Is only the log message and the same on both places.

Jalle19 commented:
Got some fresh complaints about us never fixing this bug (http://trac.kodi.tv/ticket/14498), any chance we could just show this if I fix the outstanding comments (not the architectural ones)?

Jalle19 commented:
Added some fixups, squashed the last [squash] commit.

jenkins build this please

Jalle19 commented:
Supposed to be obsolete

zag2me commented:
Is this possible to do yet? I'm using DVBS-2 and now have IPTV also, I'm happy to write a guide if it does work....

ksooo commented:
> Is this possible to do yet? I'm using DVBS-2 and now have IPTV also, I'm happy to write a guide if it does work....

 Not supported, works only by luck if at all.

 To realize your use case, I'd route iptv through tvheadend.

Jalle19 commented:
Just enable more than one PVR client

ksooo commented:
> Just enable more than one PVR client

 ... which does not work reliably as we know. No?

Jalle19 commented:
I thought it was fixed now that the manager is started asynchronously?

ksooo commented:
@FernetMenta ^^^ ?
